File: Examples/VisualBasic/IMAP/ListingMessagesWithPagingSupport.vb
Imports System.Collections.Generic
Imports Aspose.Email.Imap
Imports Aspose.Email.Mail

Namespace Aspose.Email.Examples.VisualBasic.Email.IMAP
    Class ListingMessagesWithPagingSupport
        Private Shared Sub Run()
            ' ExStart: ListingMessagesWithPagingSupport
            Using client As New ImapClient("host.domain.com", 993, "username", "password")
                Try
                    Dim messagesNum As Integer = 12
                    Dim itemsPerPage As Integer = 5
                    Dim message As MailMessage = Nothing
                    ' Create some test messages and append these to server's inbox
                    For i As Integer = 0 To messagesNum - 1
                        message = New MailMessage("from@domain.com", "to@domain.com", "EMAILNET-35157 - " + Guid.NewGuid().ToString(), "EMAILNET-35157 Move paging parameters to separate class")
                        client.AppendMessage(ImapFolderInfo.InBox, message)
                    Next

                    ' List messages from inbox
                    client.SelectFolder(ImapFolderInfo.InBox)
                    Dim totalMessageInfoCol As ImapMessageInfoCollection = client.ListMessages()
                    ' Verify the number of messages added
                    Console.WriteLine(totalMessageInfoCol.Count)

                    ' RETREIVE THE MESSAGES USING PAGING SUPPORT

                    Dim pages As New List(Of ImapPageInfo)()
                    Dim pageInfo As ImapPageInfo = client.ListMessagesByPage(itemsPerPage)
                    Console.WriteLine(pageInfo.TotalCount)
                    pages.Add(pageInfo)
                    While Not pageInfo.LastPage
                        pageInfo = client.ListMessagesByPage(pageInfo.NextPage)
                        pages.Add(pageInfo)
                    End While
                    Dim retrievedItems As Integer = 0
                    For Each folderCol As ImapPageInfo In pages
                        retrievedItems += folderCol.Items.Count
                    Next
                    Console.WriteLine(retrievedItems)
                Finally
                End Try
            End Using
            ' ExEnd: ListingMessagesWithPagingSupport
        End Sub
    End Class
End Namespace

File: Examples/CSharp/IMAP/ListingMessagesWithPagingSupport.cs
using System;
using System.Collections.Generic;
using Aspose.Email.Clients.Imap;
using Aspose.Email.Mime;

namespace Aspose.Email.Examples.CSharp.Email.IMAP
{
    class ListingMessagesWithPagingSupport
    {
        static void Run()
        {
            // ExStart: ListingMessagesWithPagingSupport
            ///<summary>
            /// This example shows the paging support of ImapClient for listing messages from the server
            /// Available in Aspose.Email for .NET 6.4.0 and onwards
            ///</summary>
            using (ImapClient client = new ImapClient("host.domain.com", 993, "username", "password"))
            {
                try
                {
                    int messagesNum = 12;
                    int itemsPerPage = 5;
                    MailMessage message = null;
                    // Create some test messages and append these to server's inbox
                    for (int i = 0; i < messagesNum; i++)
                    {
                        message = new MailMessage(
                            "from@domain.com",
                            "to@domain.com",
                            "EMAILNET-35157 - " + Guid.NewGuid(),
                            "EMAILNET-35157 Move paging parameters to separate class");
                        client.AppendMessage(ImapFolderInfo.InBox, message);
                    }

                    // List messages from inbox
                    client.SelectFolder(ImapFolderInfo.InBox);
                    ImapMessageInfoCollection totalMessageInfoCol = client.ListMessages();
                    // Verify the number of messages added
                    Console.WriteLine(totalMessageInfoCol.Count);

                    ////////////////// RETREIVE THE MESSAGES USING PAGING SUPPORT////////////////////////////////////

                    List<ImapPageInfo> pages = new List<ImapPageInfo>();
                    PageSettings pageSettings = new PageSettings();
                    ImapPageInfo pageInfo = client.ListMessagesByPage(itemsPerPage, 0, pageSettings);
                    Console.WriteLine(pageInfo.TotalCount);
                    pages.Add(pageInfo);
                    while (!pageInfo.LastPage)
                    {
                        pageInfo = client.ListMessagesByPage(itemsPerPage, pageInfo.NextPage.PageOffset, pageSettings);
                        pages.Add(pageInfo);
                    }
                    int retrievedItems = 0;
                    foreach (ImapPageInfo folderCol in pages)
                        retrievedItems += folderCol.Items.Count;
                    Console.WriteLine(retrievedItems);
                }
                finally
                {
                }
            }
            // ExEnd: ListingMessagesWithPagingSupport
        }
    }
}
